### PR TITLE
Use Airspeed Velocity for Regression Testing

### PR DESCRIPTION
**Pull request by saraedum**
### 📚 Description

…the changeset is probably not useful anymore. I just resurrected what was on https://github.com/sagemath/sage/issues/25262.

### 📝 Checklist

- [ ] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [ ] I have linked an issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### TODO

See comments in the source code. Additionally:

* [ ] Import in asv is slow for methods with a `[...]` in the name.
* [ ] Setting `docstring` as the module name fixed the import timings. However, now everything shows up as `docstring` in the benchmark grid :(

### CODE REVIEW

**Comment by mantepse** [comment]
I'd still find this wonderful!

File: src/sage/benchmark/__init__.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
+from sage.all import *
```
Comment (mezzarobba) [inline src/sage/benchmark/__init__.py]: FWIW: in #35049 (trivial PR moving some existing benchmark files to the same place) I used `sage.tests.benchmarks` instead of `sage.benchmark`. No opinion on which is best, but I guess we don't want to keep both.

**Comment by saraedum** [comment]
I ran matrix/ and rings/ for all the releases between 9.7 and 9.8.rc1. The results are https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/.

tl;dr the output is already useful but also very noisy. We should improve the regression detection to get rid of much of the noise.

If you go to regression there's a bunch of noise in the beginning (I guess) where just the rc1 was much slower than all the previous ones. So likely, there was just load on the machine when these samples were taken (maybe we can tune asv's regression detection to not detect these?)

[more noise]
<details>
The first benchmark that is not of this kind is [docstring.sage․rings․power_series_poly.track_PowerSeries_poly․truncate](https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#docstring.sage%E2%80%A4rings%E2%80%A4power_series_poly.track_PowerSeries_poly%E2%80%A4truncate?commits=047281e0-9116c558). This seems to be noise however. At least I cannot reproduce this.

The first benchmark that is not of this kind either is [docstring.sage․rings․polynomial․multi_polynomial.track_MPolynomial․is_symmetric](https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#docstring.sage%E2%80%A4rings%E2%80%A4polynomial%E2%80%A4multi_polynomial.track_MPolynomial%E2%80%A4is_symmetric?commits=54cd6fe6-6d31aad9). However, you can see that this one has recovered already in the latest rc. (Upon closer inspection, this seems to also have been noise.)
</details>

Then there's [docstring.sage․matrix․matrix0.track_Matrix․is_singular](https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#docstring.sage%E2%80%A4matrix%E2%80%A4matrix0.track_Matrix%E2%80%A4is_singular?commits=627b2bdf-4541564c) and this I can finally reproduce. So, there seems to be a speed regression here.

I can also reproduce [docstring.sage․rings․lazy_series.track_LazyPowerSeries_gcd_mixin․gcd](https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#docstring.sage%E2%80%A4rings%E2%80%A4lazy_series.track_LazyPowerSeries_gcd_mixin%E2%80%A4gcd?commits=c3028e74-84f02afa); however, since this feature was only introduced in the 9.8 series, it's likely just a bugfix that makes the computation slower.

Some things look quite striking in the graphs like [docstring.sage․rings․polynomial․polynomial_ring.track_PolynomialRing_general․_element_constructor_](https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#docstring.sage%E2%80%A4rings%E2%80%A4polynomial%E2%80%A4polynomial_ring.track_PolynomialRing_general%E2%80%A4_element_constructor_?commits=54cd6fe6-6d31aad9) but I just cannot reproduce them. While this could be noise, I doubt it somewhat. It could simply be that the surrounding tests have changed, so due to some changed caching, the timings change at some point. To support this theory somewhat, I get the slow timings on a 9.7 build.

**Comment by mantepse** [comment]
Very general question: what is actually compared, in particular, if the doctests are modified?

**Comment by saraedum** [comment]
> Very general question: what is actually compared, in particular, if the doctests are modified?

The doctests are extracted (hover over a test and you should see the extracted doctest.) When the doctest changes, the results are not related anymore. Currently, the old results are not shown at all. We could probably change this behavior and still show them but label them differently.

**Comment by mantepse** [comment]
Ah, excellent.  The doctest is only shown in the "Benchmark grid" view, that's why I missed it.

It would be quite useful to also see it in the "Regressions" view.  One can obtain it by searching for the docstring abbreviation in the docstring list, however.

In fact, it might be even better if clicking on the graph (where the regression is shown) one would get the appropriate commit in github, restricted to the file with the doctest in question.

**Comment by saraedum** [comment]
> Ah, excellent. The doctest is only shown in the "Benchmark grid" view, that's why I missed it.
> 
> It would be quite useful to also see it in the "Regressions" view. One can obtain it by searching for the docstring abbreviation in the docstring list, however.
> 
> In fact, it might be even better if clicking on the graph (where the regression is shown) one would get the appropriate commit in github, restricted to the file with the doctest in question.

I agree but I think that these are things that should be fixed in the upstream asv implementation.

**Comment by mantepse** [comment]
> > Ah, excellent. The doctest is only shown in the "Benchmark grid" view, that's why I missed it.
> > It would be quite useful to also see it in the "Regressions" view. One can obtain it by searching for the docstring abbreviation in the docstring list, however.
> > In fact, it might be even better if clicking on the graph (where the regression is shown) one would get the appropriate commit in github, restricted to the file with the doctest in question.
> 
> I agree but I think that these are things that should be fixed in the upstream asv implementation.

Since I have no clue how asv works, I feel uneasy to report this as a wishlist item.  Please feel free to do so - or not :-)

In any case, I am extremely happy that this performance tracking now seems within reach.  I can't wait to see the results for the `combinat` folder!

**Comment by mantepse** [comment]
#35127 would have been caught by this.

**Comment by saraedum** [comment]
> https://github.com/sagemath/sage/pull/35127 would have been caught by this.

I am running combinat and data_structures now on the 9.8 cycle to see if that's actually true :)

**Comment by mezzarobba** [comment]
Very nice!

> It could simply be that the surrounding tests have changed, so due to some changed caching, the timings change at some point.

Maybe recording timings for entire files as well would help with that?

> Then there's [docstring.sage․matrix․matrix0.track_Matrix․is_singular](https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#docstring.sage%E2%80%A4matrix%E2%80%A4matrix0.track_Matrix%E2%80%A4is_singular?commits=627b2bdf-4541564c) and this I can finally reproduce. So, there seems to be a speed regression here.

I can reproduce it too. I think it is due to #34484.

**Comment by saraedum** [comment]
> > #35127 would have been caught by this.
> 
> I am running (combinat and) data_structures now on the 9.8 cycle to verify that that's actually true :)

It would not have been caught directly because the doctest also changed.

**Comment by saraedum** [comment]
@mantepse if you're looking for something in the combinat namespace, this looks real:

image

**Comment by mantepse** [comment]
> > > #35127 would have been caught by this.
> > 
> > 
> > I am running (combinat and) data_structures now on the 9.8 cycle to verify that that's actually true :)
> 
> It would not have been caught directly because the doctest also changed.

No, I don't think so.  it did not change since bd3e3ebf0d7b707b9b03ab42ee7e7c71a28df782, which was merged in 9.8.beta4, and the change in `sfa.py` was introduced in  5b8a696cd07272c952bee10e1f1d4bc4546ba671 in 9.8.beta5, I think.

But I agree that it would be just luck.

**Comment by saraedum** [comment]
@mantepse I think you're right. Then something in the import went wrong. We only see 9.8.beta2 and 9.8.beta3 for that method in the asv output.

**Comment by saraedum** [comment]
@mantepse when feeding asv with previously collected data, it needs to be fed to it in commit order. I adapted my scripts. Now #35127 is detected correctly:

image

**Comment by mantepse** [comment]
That's not visible on https://www.imo.universite-paris-saclay.fr/~ruth/asv/ yet, right?

**Comment by saraedum** [comment]
> That's not visible on https://www.imo.universite-paris-saclay.fr/~ruth/asv/ yet, right?

No, it takes an hour or so to rerender everything for all the releases.

**Comment by saraedum** [comment]
@mantepse I deployed a fixed version at https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#regressions?sort=3&dir=desc

**Comment by mantepse** [comment]
Thank you!  So, all those regressions in between rc0 and rc1are still there.  That's very strange!

I guess one improvement would be to systematically have a few tests marked `# benchmark` or some such, which would test slightly larger examples and take around a second or so.  Then the noise wouldn't be a problem.

**Comment by mantepse** [comment]
As an example: https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#docstring.sage%E2%80%A4combinat%E2%80%A4decorated_permutation.track_DecoratedPermutation%E2%80%A4to_signed_permutation shows a clear increase, but it is completely fake.

**Comment by mantepse** [comment]
I just realised that sorting only works for the "factor" column.

**Comment by saraedum** [comment]
> As an example: https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#docstring.sage%E2%80%A4combinat%E2%80%A4decorated_permutation.track_DecoratedPermutation%E2%80%A4to_signed_permutation shows a clear increase, but it is completely fake.

I agree. I think the problem was that while testing the last few betas and rcs there was additional load on the system that we start to see in some tests.

The hope here would be that having a much finer iteration of benchmark runs (we test with every commit to develop, we just need to export the results) such noise would not be able to create such misleading shapes.

**Comment by saraedum** [comment]
Let me run things for 9.8. Let's see how that changes the pictures.

**Comment by mezzarobba** [comment]
> I agree. I think the problem was that while testing the last few betas and rcs there was additional load on the system that we start to see in some tests.

How hard would it be to aggregate the timings from several runs (and detect regressions based on the minimum run time of the test for each revision, say)? From a quick look at the asv docs, it is not clear to me if this is already supported in one way or another: can we just we feed it data from several benchmark runs on the same machine and trust it to do something sensible with it? Are regression detected separately for each machine or is there a notion of a “global” regression?

Also, is it useful if we try to reproduce your experiments on a different system to have a second machine in the results? How did you run the code with older versions of Sage, given that it does not rebase cleanly? (I suppose `git rebase -X ours` should suffice?...)

**Comment by saraedum** [comment]
> How hard would it be to aggregate the timings from several runs (and detect regressions based on the minimum run time of the test for each revision, say)?

That would be very easy. But the idea here is to use the timings from the test run in GitHub's CI where we only want to run all the tests once.

**Comment by saraedum** [comment]
> Also, is it useful if we try to reproduce your experiments on a different system to have a second machine in the results? How did you run the code with older versions of Sage, given that it does not rebase cleanly? (I suppose git rebase -X ours should suffice?...)

I don't think that's very useful. We are going to have similar (or worse) amounts of noise when we use GitHub CI instead. So what we come up with should work for a very noisy runner.

I just extracted `develop...asv` into a patch file and it applied cleanly to all these versions that I built from source.

**Comment by saraedum** [comment]
> Let me run things for 9.8. Let's see how that changes the pictures.

Same picture:

image

@mantepse And actually, I can reproduce this regression. I paste the following in a sage repl (the preceding doctests)

```
DecoratedPermutation([2, 1, 3])
DecoratedPermutation([2, 1, -3])
S = DecoratedPermutations(3)
elt = S([2, 1, -3])
TestSuite(elt).run()
S = DecoratedPermutations(3)
elt = S([2, 1, -3])
elt.check()
elt = S([2, -1, 3])
len(set(hash(pi) for pi in DecoratedPermutations(3)))
S = DecoratedPermutations(3)
elt1 = S([2, 1, -3])
elt2 = DecoratedPermutation([2, 1, -3])
elt1 == elt2
S = DecoratedPermutations(3)
elt1 = S([2, 1, -3])
elt2 = DecoratedPermutation([2, 1, 3])
elt1 != elt2
DecoratedPermutation([2, 1, -3]).size()
```

and then I call

```
reset()
%time DecoratedPermutation([2, 1, -3]).to_signed_permutation()
Wall time: 3.64 ms  # 9.7
Wall time: 11.3 ms  # 10.0.beta0
```

**Comment by saraedum** [comment]
So, one conclusion here could be that we need a little script to run all doctests up to some line to make it easier to "debug" these kinds of reports.

**Comment by saraedum** [comment]
Without any tricks, I can reproduce this slowdown (since @mantepse probably cares about species?)

image

**Comment by saraedum** [comment]
I tweaked the regression detection data a bit and updated the deployed demo at https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#regressions?sort=3&dir=desc

Now the detected major regressions look legit to me (at least the graphs on the first page are mostly convincing.)

**Comment by mantepse** [comment]
> @mantepse And actually, I can reproduce this regression. I paste the following in a sage repl (the preceding doctests)
> ...
> `%time DecoratedPermutation([2, 1, -3]).to_signed_permutation()`

Hm, on my box (ubuntu 22.04.1, but I guess that's irrelevant), I can only reproduce a slight difference in a fresh session, I get about 2.6ms vs 3.45ms.

If I do the test (including the stuff that came before) again, the difference is gone.  What puzzles me, however, is that this regression (if it is real) should be visible also in other doctests.

Still, I'm not sure how relevant this is.  I really think it would be better to have a few specially marked tests (i.e., `# benchmark` which take slightly longer, and only for some exposed methods.  I guess that this wouldn't change the framework a lot, would it?  Do you currently run the long doctests, too?

In the case at hand, for example:
```
sage: l = DecoratedPermutations(8)
sage: _ = [pi.to_signed_permutation() for pi in l]

sage: set_random_seed(1783)
sage: pi = DecoratedPermutations(100).random_element()
sage: pi.to_signed_permutation()
```
(and this won't work, because it exhibits the missing random generator, which is #34925)

**Comment by mezzarobba** [comment]
> > How hard would it be to aggregate the timings from several runs (and detect regressions based on the minimum run time of the test for each revision, say)?
> 
> That would be very easy. But the idea here is to use the timings from the test run in GitHub's CI where we only want to run all the tests once.

Yes, I was thinking more about older releases.

**Comment by saraedum** [comment]
@mantepse did you run the test as I had posted it, i.e., with the `reset` that happens between doctests? Without it, I also don't get the regression.

**Comment by saraedum** [comment]
> Still, I'm not sure how relevant this is. I really think it would be better to have a few specially marked tests (i.e., # benchmark which take slightly longer, and only for some exposed methods.

For benchmarking explicitly, I propose to use asv the way it is meant to be used, i.e., just drop some benchmarks in the `src/sage/benchmark/` directory, see https://asv.readthedocs.io/en/stable/writing_benchmarks.html#writing-benchmarks

> I guess that this wouldn't change the framework a lot, would it? Do you currently run the long doctests, too?

No, I only run the normal doctests. This is also what is being run by the GitLab CI on each merge currently.

**Comment by mantepse** [comment]
> > Still, I'm not sure how relevant this is. I really think it would be better to have a few specially marked tests (i.e., # benchmark which take slightly longer, and only for some exposed methods.
> 
> For benchmarking explicitly, I propose to use asv the way it is meant to be used, i.e., just drop some benchmarks in the `src/sage/benchmark/` directory, see https://asv.readthedocs.io/en/stable/writing_benchmarks.html#writing-benchmarks

The problem is that this disconnects code from tests.  I would guess that we would be much more encouraged to write benchmark tests, if they are kept in the same source.

Possibly one could also automate some for certain categories, but I think it would be great if we could start somehow and refine later.

If benchmark tests take around 1 second, I suspect that noise would also be less of an issue.

> > I guess that this wouldn't change the framework a lot, would it? Do you currently run the long doctests, too?

> No, I only run the normal doctests. This is also what is being run by the GitLab CI on each merge currently.

Did you try to run the long doctests, or do they take too long?

More explicitly, I propose to provide infrastructure for either a new section `BENCHMARKS::`, or a new marker `# benchmark`, which is ignored for testing.

I think we could specify that these tests should be independent of other tests and should, if at all possible, never be modified.

What do you think?

**Comment by saraedum** [comment]
> What do you think?

I am a bit torn here but I think we should try to just use the tools in the way they are meant to be used. We have some benchmarks in SageMath already and nobody cares about them it seems, so that shows that we should maybe not create a `sage.benchmark` namespace. And using the doctest approach proposed in this PR, we are actually not using asv as it is meant to be used. But I guess it just feels like the minimal amount of work necessary to get some benchmarks at least while not creating lots of logic for benchmark filtering that needs to be maintained.

A marker `# benchmark` would indeed be easy to implement. A marker `BENCHMARKS::` is much more work I think. If we even call it `# optional: benchmark` then it's quite easy I think. It is, however, not clear to me how these benchmarks get executed? Do the lines above that do not say `# benchmark` also get executed first? In any case, the downside is that we need to write documentation and maintain this feature. So maybe it's just easier to say: all our doctests are benchmarks (long or not, optional or not, depending on how our CI is configured) and for proper benchmarks, please read the asv documentation.

**Comment by mantepse** [comment]
I am mostly interested in avoiding performance regressions, because I spent quite some time fixing constructors that were unnecessarily slow.  In particular, these were an obstacle for findstat, some, for example the constructor for posets, still are.

Thus, I would love to see a bot that runs on every ticket, or at least on every beta relaease, which reliably warns us of (major) regressions.

Is this possible with the current approach?  Are the timings always taken on the same machine, or is there some magic that accounts for different machines?
 
If we can run the long doctests on github, then maybe we do not need an extra marker - just mention in the developer's guide that these are used to detect performance regressions.  However, I'm afraid that we will hit the problem that we cannot compare doctests which changed.  Do I understand correctly that doctests for a method cannot be compared if even one of them was modified?

Here is yet another idea to avoid the problem of changing doctests and to get started more quickly.  We could have a marker `# optional: benchmark XX.X[.betaX]`, with the following meaning:

* for every method *only* doctests marked such are run and measured.
* `XX.X[.betaX]` specifies the first version number for which the doctest should be run (most importantly, any version number for which the test passes)

Then (if this is possible at all), the bot would run this test with all binaries available which are younger than `XX.X[.betaX]` and additionally with the patch proposed in the ticket.

A prerequisite is, of course, that we can access pre-built binaries.  Is this possible - difficult - easy?

To get started, we could add this marker to the `long` doctests for which it makes sense.

This approach might have some advantages:

* timings would be taken on the same machine
* by focussing on doctests which do not just take a few milliseconds we might have less noise
* we would avoid the problem of changing doctests

If this approach is easy to implement, there are some later optimizations possible:

* we might be able to store timings taken on a particular machine for a particular doctest.
* if this is not possible, I guess that most of the time, it would only be interesting to compare with the betas of the current release, so we could avoid running the tests on old releases.

**Comment by saraedum** [comment]
> Thus, I would love to see a bot that runs on every ticket, or at least on every beta relaease, which reliably warns us of (major) regressions.
> Is this possible with the current approach? Are the timings always taken on the same machine, or is there some magic that accounts for different machines?

I don't think this is possible reliably. Other projects have experimented with this in the past and from the examples I know are not using this setup anymore. Even if you run the benchmarks on the same machine every time, you need to go to some length to get reproducible timings (disable power management, make sure there's nobody else on the same hardware, so no virtual machines, make sure nothing else runs in parallel, so no parallel execution of tests, …)

In principle the timings are on very similar machines. The tests run on GitHub CI virtual machines that have in my experience very similar hardware. If the hardware changes a lot, one can detect the generation of the machine and set a `machine` in asv to separate runs on different hardware. (There's no magic to relate timings on different hardware, they are treated as independent data.)

However, to your `Is this possible?` question. You could add actual benchmarks that test the constructors properly to `sage.benchmark`. You can configure these to run the constructors many times so you get meaningful timings. The asv website publishes an rss feed of regressions that you could subscribe to and filter for the things that you care about. Then you would get notified about regressions in your benchmarks with every merge into the develop branch. You can then verify these regressions and do something about them, e.g., complain with the PR that broke things. (I strongly advise against automating this. The project I know removed this again after ignoring it for a while because it was way too noisy.)

> Do I understand correctly that doctests for a method cannot be compared if even one of them was modified?

Yes if one line of the the actual doctested content changes, the old timings are not shown anymore. I think this is the correct approach. I don't want to use doctests of actively developed code for benchmarking. We should write proper benchmarks for this. But I want to see that much of number theory got a bit slower because somebody changed a small detail of how `right_kernel` works ;)

> Here is yet another idea to avoid the problem of changing doctests [...]

Sorry, this is not easily possible. We don't even have all of the recent releases available as docker images or conda builds and they even have dependencies moving out of sync with how the SageMath distribution has updated them in some cases. We have no recent betas available as binaries in that format.

**Comment by saraedum** [comment]
> > I don't think that's very useful. We are going to have similar (or worse) amounts of noise when we use GitHub CI instead. So what we come up with should work for a very noisy runner.
> 
> Same here.

@mezzarobba do you think that the regressions shown at https://www.imo.universite-paris-saclay.fr/~ruth/asv/html/#regressions?sort=3&dir=desc now look reasonable enough?

**Comment by saraedum** [comment]
Yes, I only ran combinat. Let me run all of SageMath. I'll ping you tomorrow with the full output :)

**Comment by mantepse** [comment]
Sorry for asking so many questions: if it is not possible to reliably detect major regressions, what is the aim of this ticket?

**Comment by saraedum** [comment]
I think we can reliably detect major regressions a while after they happened and we can then also detect where they happened. We can also get good hints at where a regression might have happened. For actual benchmarks in `sage.benchmark`, we get a tool to run these benchmarks during development locally and time things properly without having to manually resort to `%%timeit`.

**Comment by mantepse** [comment]
Concerning

> Sorry, this is not easily possible. We don't even have all of the recent releases available as docker images or conda builds and they even have dependencies moving out of sync with how the SageMath distribution has updated them in some cases. We have no recent betas available as binaries in that format.

and

> I think we can reliably detect major regressions a while after they happened and we can then also detect where they happened. We can also get good hints at where a regression might have happened. For actual benchmarks in `sage.benchmark`, we get a tool to run these benchmarks during development locally and time things properly without having to manually resort to `%%timeit`.

I don't really understand.

1. Why would it be hard to create binaries for the last few releases and betas?  I admit, I don't know how much space a binary takes, is this the problem?  To make benchmarking as proposed interesting, I think that having binaries for a few releases after sage 9.0 and for the last betas until the current one would suffice completely.  If some are missing (because of dependency problems) that wouldn't really hurt either, would it?
2. What is `sage.benchmark` - is this a proposal, or something I am missing?  In any case, what I believe is most important is that benchmarks are not separated from other code.  The mental barrier to write a simple benchmark in the same file is much lower.

Having such a benchmarking tool as a bot would be nice, but I don't think it is absolutely necessary.  In fact, I think it would even be sufficient if it is run for every beta, and not for every ticket, if that's a problem.

If it is less work for you: maybe it would be more efficient to have a video conference, so you could explain the problems to me in person?

**Comment by saraedum** [comment]
> Why would it be hard to create binaries for the last few releases and betas?

We have failed to find a maintainer for the docker images for the past few years, i.e., have somebody build binaries and make sure that the script used to build those binaries works somewhat reliably on some sort of CI hardware. (Part of the story was also that we couldn't find anybody to maintain that CI infrastructure …)

I am very reluctant to create any process that does become more than just a tiny added step in our current setup because it creates a burden to maintain. (We run the CI on the doctests anyway, so let's just export the data it collects. And while we parse that data, asv will also run the benchmarks that are explicitly marked as such.)

> What is sage.benchmark?

Sorry, if that's not clear. With the changes in this PR, asv runs all benchmarks that are in `sage.benchmark`. Currently, there's only the "benchmark" that parses the output from a previous doctest run and reports those timings to asv. We should for a start also convert the existing benchmarks (somewhat spread out in the special benchmark files that probably no one is aware of...) to be asv benchmarks that reside in `sage.benchmark`.

If you want to write benchmarks that are implemented outside of `sage.benchmark`, you can just add a benchmark class to your module and make sure it gets imported in some way by say `sage.benchmark.all` or `sage.benchmark.combinat.all` or whatever feels most natural over time.

> If it is less work for you: maybe it would be more efficient to have a video conference, so you could explain the problems to me in person?

Sure. I can do today at 6pm Germany time or basically any time tomorrow. Let's see if we can agree on some time at https://sagemath.zulipchat.com/#narrow/stream/271079-doctest/topic/tracking.20performance.20regressions

**Comment by mantepse** [comment]
Any news here?

**Comment by saraedum** [comment]
No, I had no time to work on this.

**Comment by mantepse** [comment]
> No, I had no time to work on this. (And had missed your question.)

Looking forward to the invention of the time turner :-)
